### PR TITLE
An initial implementation of graph query evaluation

### PR DESCRIPTION
**Pull request by vgapeyev**
An end-to-end implementation of graph pattern matching that can be unit-tested and experimented with in REPL. 

## Relevant Issues
In combination with a few earlier PRs, this is meant to close the trio of issues #996, #997, #998. 

## Description

The targeted language definition is what was proposed in [RFC-0025](https://github.com/partiql/partiql-docs/blob/main/RFCs/0025-graph-data-model.md) and [RFC-0033](https://github.com/partiql/partiql-docs/blob/main/RFCs/0033-graph-query.md).

The fragment of GPML patterns that is implemented can be characterized as: straight-path patterns with simple label matching and all directed/undirected edge patterns. 
This already allows one to write some interesting queries can be written and tested already.  Some examples are in the unit tests `EvaluatingCompilerGraphMatchTests`.

The design of neither the internal graph representation nor the evaluation engine is nowhere settled and is meant to evolve as more GPML features are implemented and more is learned. 

The evaluation approach is meant to follow the one outlined in Section 6 of the GPML paper [[ACM](https://dl.acm.org/doi/10.1145/3514221.3526057), [arxiv](https://arxiv.org/abs/2112.06217).]
In particular, the computational core of this initial implementation corresponds to Section 6.4, Computation of Path Binding, which deals with elementary "step" patterns and joining their results.  

The most interesting parts are probably the graph abstraction `Graph.kt` and some ideas for graph pattern "plans" in `MatchSpec.kt`, for structuring the matching algorithms in `GraphEngine.kt`.  

## Other Information
- Updated Unreleased Section in CHANGELOG: YES
- Any backward-incompatible changes? NO
- Any new external dependencies? NO

## License Information

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/SimpleGraph.kt [hunk 1]
```diff
@@ -32,4 +32,57 @@ class SimpleGraph(
...
+    private fun labelsMatchSpec(labels: Set<String>, spec: LabelSpec): Boolean =
+        when (spec) {
+            LabelSpec.Whatever -> true
+            is LabelSpec.OneOf -> labels.contains(spec.name)
+        }
+
+    private fun Graph.Elem.matches(labelSpec: LabelSpec): Boolean =
+        labelsMatchSpec(this.labels, labelSpec)
+
+    override fun getNodes(spec: LabelSpec): List<Node> {
```
Comment (jpschorr) [inline partiql-lang/src/main/kotlin/org/partiql/lang/graph/SimpleGraph.kt:45]: maybe a name that implies searching/filtering; `getNodes` implied returning all nodes to me.

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+        val plan = planStride(stride)
+        check(stride == restoreStrideSpec(plan))
+        return evaluatePlan(graph, plan)
+    }
+
+    // This is for a node-only pattern like  (g MATCH (x)). TODO: Something less ugly / more unified with the rest?
+    private fun evaluateNodeStride(graph: Graph, stride: StrideSpec): StrideResult {
+        check(stride.elems.size == 1)
+        return when (val node = stride.elems[0]) {
+            is EdgeSpec -> { check(false); /*dummy*/ StrideResult(stride, emptyList()) }
```
Comment (jpschorr) [inline partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:42]: Is this `check` and `dummy` intended to essentially assert that this branch should be unreachable?

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+    fun evaluate(graph: Graph, matchSpec: MatchSpec): MatchResult {
+        val strideResults = matchSpec.strides.map { evaluateStride(graph, it) }
+        return joinStridesOnBinders(strideResults)
+    }
+
+    private fun evaluateStride(graph: Graph, stride: StrideSpec): StrideResult {
+        if (stride.elems.size == 1) return evaluateNodeStride(graph, stride)
+
+        val plan = planStride(stride)
+        check(stride == restoreStrideSpec(plan))
```
Comment (jpschorr) [inline partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:34]: This restoration is essentially a test that the Spec is round-trippable to plan as a sanity check?

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        }
+    }
+
+    /** Determine the order of joining the steps in a stride.
+     *  Assumes that [stride] contains a properly alternating list: n, e, n, ..., e, n.
+     *
+     *  At the moment, this just takes one straightforward order.
+     *  Conceivably, this can become something more sophisticated.
+     */
+    private fun planStride(stride: StrideSpec): StrideTree {
```
Comment (jpschorr) [inline partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:56]: Should this planning be lifted into GPMLTranslator? This seems to perhaps be a preliminary for of the `expansion` form section 6.3 of the GPML paper?

If the planning were done by GPML translator, then any errors encountered could be thrown prior to the start of evaulation.

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt [hunk 5]
```diff
@@ -0,0 +1,160 @@
...
+        fun leafFrom3(elems: List<ElemSpec>): StrideLeaf {
+            check(elems[0] is NodeSpec)
+            check(elems[1] is EdgeSpec)
+            check(elems[2] is NodeSpec)
+            return StrideLeaf(StrideSpec(elems.take(3)))
+        }
+
+        fun planRightLeaning(elems: List<ElemSpec>): StrideTree =
+            when (elems.size) {
+                0, 1, 2 -> { check(false); /*dummy*/ StrideLeaf(StrideSpec(emptyList())) }
```
Comment (jpschorr) [inline partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:68]: Is this check and dummy intended to essentially assert that this branch should be unreachable?

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt [hunk 6]
```diff
@@ -0,0 +1,160 @@
...
+        )
+    }
+
+    /** Joins results of stride matches on distinct path patterns of a graph pattern.
+     *  In general, this is a cartesian product, but it is whittled down by the requirement
+     *  that a given binder variable binds to the same graph element in each individual answer.
+     */
+    fun joinStridesOnBinders(strides: List<StrideResult>): MatchResult {
+        return when (strides.size) {
+            0 -> { check(false); return /*dummy*/ MatchResult(emptyList(), emptyList()) }
```
Comment (jpschorr) [inline partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:149]: Is this check and dummy intended to essentially assert that this branch should be unreachable?

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GpmlTranslator.kt [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+    /** The main entry point into the translator. */
+    fun translateGpmlPattern(gpml: PartiqlAst.GpmlPattern): MatchSpec {
+
+        if (gpml.selector != null) TODO("Evaluation of GPML selectors is not yet supported")
+        return MatchSpec(gpml.patterns.map { StrideSpec(patchElemList(translatePathPat(it))) })
+    }
+
+    fun translatePathPat(path: PartiqlAst.GraphMatchPattern): List<ElemSpec> {
+        if (path.prefilter != null || path.quantifier != null || path.restrictor != null || path.variable != null)
+            TODO("Not yet supported in evaluating a GPML path pattern: prefiletrs, quantifiers, restrictors, binder variables.")
```
[the remark this answers is not in the document]
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GpmlTranslator.kt]: Fixed on my side.

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GpmlTranslator.kt [hunk 8]
```diff
@@ -0,0 +1,99 @@
...
+            is PartiqlAst.GraphMatchDirection.EdgeLeftOrRight -> DirSpec.`(-)`
+            is PartiqlAst.GraphMatchDirection.EdgeLeftOrUndirectedOrRight -> DirSpec.`---`
+        }
+
+    /** Make sure there is proper alternation of NodeSpec and EdgeSpec entries,
+     *  by inserting a [NodeSpec] between adjacent [EdgeSpec]s.
+     *  TODO: Deal with adjacent [NodeSpec]s -- by "unification" or prohibit.
+     */
+    fun patchElemList(elems: List<ElemSpec>): List<ElemSpec> {
+        // println("Before patching: ${elems}")
```
[the remark this answers is not in the document]
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GpmlTranslator.kt]: removed

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GpmlTranslator.kt [hunk 9]
```diff
@@ -0,0 +1,99 @@
...
+                }
+            } else { // expectNode == false
+                when (x) {
+                    is NodeSpec -> TODO("Deal with adjacent nodes in a pattern.  Unify? Prohibit?")
+                    is EdgeSpec -> { patched.add(x); expectNode = true }
+                }
+            }
+        }
+        if (expectNode) patched.add(fillerNode)
+        // println("After  patching: ${patched}")
```
[the remark this answers is not in the document]
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GpmlTranslator.kt]: removed

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt (hunk 3 above)
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:34]: yes.  I'd like to keep the checks in while it's still actively developed.

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt (hunk 2 above)
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:42]: yes

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt (hunk 5 above)
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:68]: yes

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt (hunk 6 above)
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/GraphEngine.kt:149]: yes

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/SimpleGraph.kt (hunk 1 above)
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/SimpleGraph.kt:45]: `scanNodes` ?

File: partiql-lang/src/test/kotlin/org/partiql/lang/eval/GraphsInSessionTests.kt [hunk 10]
```diff
@@ -0,0 +1,146 @@
...
+        mapOf(
+            "aG1" to ExprValue.newGraph(ApiSampleGraphs.g1),
+            "aG1L" to ExprValue.newGraph(ApiSampleGraphs.g1L),
+            "aG2" to ExprValue.newGraph(ApiSampleGraphs.g2),
+            "aG3" to ExprValue.newGraph(ApiSampleGraphs.g3),
+
+            "tG1" to graphOfText(TextSampleGraphs.g1),
+            "tG1L" to graphOfText(TextSampleGraphs.g1),
+            "tG2" to graphOfText(TextSampleGraphs.g1),
+            "tG3" to graphOfText(TextSampleGraphs.g1),
```
[the remark this answers is not in the document]
Comment (vgapeyev) [reply partiql-lang/src/test/kotlin/org/partiql/lang/eval/GraphsInSessionTests.kt]: yes, thank you!

File: partiql-lang/src/main/kotlin/org/partiql/lang/eval/EvaluatingCompiler.kt [hunk 11]
```diff
@@ -1646,6 +1653,58 @@ internal class EvaluatingCompiler(
...
+        val m = mutableMapOf<String, Graph.Elem>()
+        for ((s, e) in elemSpecs zip elems) {
+            check((s is NodeSpec && e is Graph.Node) || (s is EdgeSpec && e is Graph.Edge))
+            // Populate the map for each binder variable in the spec,
+            // while checking that if a variable is met more than once, it is bound to the same element.
+            // Note: only singleton variables are currently supported. (No group variables yet.)
+            if (s.binder != null) {
+                val v = s.binder!!
+                if (m.containsKey(v)) {
+                    check(m[v] == e)
```
[the remark this answers is not in the document]
Comment (jpschorr) [reply partiql-lang/src/main/kotlin/org/partiql/lang/eval/EvaluatingCompiler.kt:1693]: In that case, this should probably be something more user-facing than a simple `check` call.  It's probably all right if we don't know what that should be yet. But if both 'sanity check' and 'user check' assertions are done with just `check` it'll be difficult to come back later and find and update the 'user check' sites.

**Review by vgapeyev — COMMENTED** [review]
Josh, I believe I have addressed all the comments now.

**Comment by jpschorr** [comment]
> ### Conformance comparison report
> 	Base ([33f907f](https://github.com/partiql/partiql-lang-kotlin/commit/33f907ff8ca2e7217425e19c1e93b062d2c73e7e)) 	[61c5e13](https://github.com/partiql/partiql-lang-kotlin/commit/61c5e1308ec2f0d2456487e6dccf38883098377d) 	+/-
> % Passing 	97.42% 	96.99% 	-0.43%
> ✅ Passing 	4271 	4347 	76
> ❌ Failing 	113 	135 	22
> 🔶 Ignored 	0 	0 	0
> Total Tests 	4384 	4482 	98
> 
> Number passing in both: 4269
> 
> Number failing in both: 113
> 
> Number passing in Base ([33f907f](https://github.com/partiql/partiql-lang-kotlin/commit/33f907ff8ca2e7217425e19c1e93b062d2c73e7e)) but now fail: 0
> 
> Number failing in Base ([33f907f](https://github.com/partiql/partiql-lang-kotlin/commit/33f907ff8ca2e7217425e19c1e93b062d2c73e7e)) but now pass: 0

Something is not quite right with the Conformance reporting it seems, as the table doesn't match the statistics printed after it....

Either way, it would be good to verify that all the failures are due to the fact that this PR brings in updates to partiql-tests

File: partiql-lang/src/main/kotlin/org/partiql/lang/graph/MatchSpec.kt [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+// This is yet to be thought out.
+
+/** Specification of an edge direction.
+ *  The boolean fields at the enum values capture their semantics in a way that is useful for
+ *  "compiling" them to graph scans.
+ */
+enum class DirSpec(val wantLeft: Boolean, val wantUndir: Boolean, val wantRight: Boolean) {
+    // Attempting to be close to the GPML notation, for mnemonics.
+    // However characters <, > are not supported in identifiers, so it is `--)` instead of `->`, etc.
+    // The mnemonics attempt might be futile: currently these are only referred to
```
[the remark this answers is not in the document]
Comment (jpschorr) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/MatchSpec.kt:15]: If you want same length, best I can think of is (though not sure I love it either):
```
DirL__
Dir_U_
Dir__R
DirLU_
Dir_UR
DirL_R
DirLUR
```

[the remark this answers is not in the document]
Comment (vgapeyev) [reply partiql-lang/src/main/kotlin/org/partiql/lang/graph/MatchSpec.kt:15]: Looks good to me, thanks!

**Comment by vgapeyev** [comment]
The matter with the conformance tests might be due to an inadvertently skipped update of the Git submodule during one of the merges. :-( 
At this point it seems prudent to wait until the upstream #1073 (which does not have this issue) merged and seeing then what to do.

**Comment by vgapeyev** [comment]
This branch is now back to using the same commit of the conformance tests that `main` uses.  
Tinkering with the conformance tests was not yet the intent.

**Review by jpschorr — APPROVED** [review]

**Pull request merged**